### PR TITLE
Enumerate Drives on Linux

### PR DESCRIPTION
**Pull request by minhvirtium**
- Enumerate Drives which plug in host
- Drives is devices which has type is "Disk"
- Get Drives and add to collection in order to manage.

### CODE REVIEW

File: vtStor/DriveManager.cpp [hunk 1]
```diff
@@ -62,6 +62,8 @@ eErrorCode cDriveManager::EnumerateDrives( eScanForHardwareChanges ScanForHardwa
     return( error );
 }
 
+
```
Comment (dientc) [inline vtStor/DriveManager.cpp]: Remove break line if not necessary

Comment (minhvirtium) [inline vtStor/DriveManager.cpp]: I fixed

File: StorageUtility/Linux/Device.cpp [hunk 2]
```diff
@@ -21,84 +21,61 @@ limitations under the License.
...
-/*void cDevice::Data(std::unordered_map<eDeviceDataType, void*>& Data)
-{
-  //  Data.insert({ eDeviceDataType::PSP_DEVINFO_DATA, m_DevInfoData });
-   // Data.insert({ eDeviceDataType::PSP_DEVICE_INTERFACE_DATA, m_DevInterfaceData });
-   // Data.insert({ eDeviceDataType::PSP_INTERFACE_DEVICE_DETAIL_DATA, m_DevDetailData });
-}*/
+}
 
-/*DeviceHandle cDevice::Handle()
+cDevice::cDevice (String Path)
```
Comment (tranminhtam) [inline StorageUtility/Linux/Device.cpp]: - Why cDevice      (Sring...)
Use ref bypassed

File: StorageUtility/Linux/Device.h [hunk 3]
```diff
@@ -33,29 +29,24 @@ namespace vtStor
...
-
-    void AllocateMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData);
-
-    void CopyMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData);
+    virtual void Data(std::unordered_map<eDeviceDataType, void*>& Data) override;
+    virtual DeviceHandle Handle() override;
+    virtual void DevicePath(tchar*& DevicePath) override;
 
-    void DeallocateMemories();*/
+    String GetDevicePath();
```
Comment (tamvirtium) [inline StorageUtility/Linux/Device.h:41]: Why define more function get and set for DevicePath and DeviceHandle?
Don't we have the Handle() and DevicePath(tchar*& ...) already

File: StorageUtility/Linux/Device.h [hunk 4]
```diff
@@ -33,29 +29,24 @@ namespace vtStor
 class cDevice : public cDeviceInterface
 {
 public:
-//    cDevice(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData);
-//    virtual ~cDevice();
+    cDevice();
+    cDevice (String Path);
```
Comment (tamvirtium) [inline StorageUtility/Linux/Device.h:33]: - Why cDevice      (String...)
  Use ref bypassed

File: StorageUtility/Linux/StorageUtility.cpp [hunk 5]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
 #include <string.h>
+#include <unistd.h>
 
-#include "StorageUtility.h"
 #include "Device.h"
+#include "StorageUtility.h"
 
 namespace vtStor
 {
+std::map<String, vtStor::sStorageAdapterProperty> s_MappingSystemPathToProperties;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:32]: We should not use the static to store something here

File: StorageUtility/Linux/StorageUtility.cpp [hunk 6]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
 
 eErrorCode GetStorageDevices(std::vector<std::shared_ptr<cDeviceInterface>>& Devices, eOnErrorBehavior OnErrorBehavior)
 {
-    eErrorCode error = eErrorCode::None;
-
-    sEnumerateDevicesCallback callBack;
-    callBack.Data = (void*)&Devices;
+    std::vector<String> Paths;
+    vtStor::GetStorageDevicePaths(Paths, eOnErrorBehavior::Continue);
+    for (int i = 0; i < Paths.size(); i++)
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:39]: should be ++i
please update for the remains

File: StorageUtility/Linux/StorageUtility.cpp [hunk 7]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+std::map<String, vtStor::sStorageAdapterProperty> s_MappingSystemPathToProperties;
+std::map<DeviceHandle, vtStor::String>  s_MappingDeviceHandleToSysPath;
 
 eErrorCode GetStorageDevices(std::vector<std::shared_ptr<cDeviceInterface>>& Devices, eOnErrorBehavior OnErrorBehavior)
 {
-    eErrorCode error = eErrorCode::None;
-
-    sEnumerateDevicesCallback callBack;
-    callBack.Data = (void*)&Devices;
+    std::vector<String> Paths;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:37]: Name convention

File: StorageUtility/Linux/StorageUtility.cpp [hunk 8]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+        std::shared_ptr<vtStor::cDevice> Device = std::make_shared<cDevice>(Paths[i]);
+        Devices.push_back(Device);
+    }
 
-//    callBack.Function = [](void* Data, const HDEVINFO& DevInfoHandle, SP_DEVINFO_DATA& DevInfoData, SP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData, eErrorCode& ErrorCode)
-//    {
-//        std::vector<std::shared_ptr<vtStor::cDeviceInterface>>* devices = (std::vector<std::shared_ptr<vtStor::cDeviceInterface>>*)Data;
-//        devices->push_back(std::make_shared<cDevice>((SP_DEVINFO_DATA*)&DevInfoData, (SP_DEVICE_INTERFACE_DATA*)&DevInterfaceData, DevDetailData, SizeOfDevDetailData));
-//    };
+    eErrorCode error = eErrorCode::None;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:45]: return directly to eErrorCode::None instead

File: StorageUtility/Linux/StorageUtility.cpp [hunk 9]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-                        sizeof(STORAGE_DESCRIPTOR_HEADER),
-                        &bytesReturned,
-                        NULL
-                        );
-    if ( 0 == ret )
+    if ((Handle = open(DevicePath.c_str(), O_RDONLY|O_NONBLOCK)) < 0)
     {
-        return( eErrorCode::Io );
+        printf("ERROR: Cannot open device %s\n", DevicePath.c_str());
+        return eErrorCode::None;
```
Comment (hai-tran) [inline StorageUtility/Linux/StorageUtility.cpp:155]: Please don't return error none here. ErrorNone means the operation is ok, so the out side will think "I get the DeviceHandle for this device path".

File: StorageUtility/Linux/StorageUtility.cpp [hunk 10]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
 }
 
-
-/*
-eErrorCode GetStorageDevicePaths( std::vector<String>& Paths, eOnErrorBehavior OnErrorBehavior )
+bool IsStorageDisk(udev_device* Dev)
 {
-    return( GetDevicePaths( Paths, &GUID_DEVINTERFACE_DISK, OnErrorBehavior ) );
+    const char* devt = udev_device_get_devtype(Dev);
+    if(0 == strcmp(devt, "disk"))
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:53]: code convention

File: StorageUtility/Linux/StorageUtility.cpp [hunk 11]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
     return(error);
 }
 
-
-/*
-eErrorCode GetStorageDevicePaths( std::vector<String>& Paths, eOnErrorBehavior OnErrorBehavior )
+bool IsStorageDisk(udev_device* Dev)
 {
-    return( GetDevicePaths( Paths, &GUID_DEVINTERFACE_DISK, OnErrorBehavior ) );
+    const char* devt = udev_device_get_devtype(Dev);
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:52]: do not shorten the name too much

File: StorageUtility/Linux/StorageUtility.cpp [hunk 12]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-//    if (FALSE == CloseHandle(Handle))
-//    {
-//        //! TODO
-//        //fprintf(stderr, "\nCloseDeviceHandle was not successful. Error Code: %d", GetLastError());
-//    }
-//    Handle = INVALID_HANDLE_VALUE;
-}
-
-bool IsAtaDeviceBus(udev_device* Dev)
+eErrorCode ShowDevicePaths(const std::vector<String> Paths)
```
Comment (hai-tran) [inline StorageUtility/Linux/StorageUtility.cpp:178]: Debug function?

File: StorageUtility/Linux/StorageUtility.cpp [hunk 13]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+    return false;
 }
-*/
 
-/*eErrorCode GetDevicePaths( std::vector<String>& Paths, const GUID* InterfaceClassGUID, eOnErrorBehavior OnErrorBehavior )
+bool IsAtaDeviceBus(udev_device* Dev)
 {
-    eErrorCode error = eErrorCode::None;
+    const char* bus = udev_device_get_property_value(Dev, "ID_BUS");
+    if(NULL == bus)
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:63]: Use null_ptr instead
Code convention

File: StorageUtility/Linux/StorageUtility.cpp [hunk 14]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-    case BusTypeSata:
-    {
-        return( true );
-    } break;
-    default:
-    {
-        return( false );
-    } break;0
-    }*/
+    // TODO:
```
Comment (hai-tran) [inline StorageUtility/Linux/StorageUtility.cpp:189]: Please add more comments to describe what is the TODO here

File: StorageUtility/Linux/StorageUtility.cpp [hunk 15]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+    const char* bus = udev_device_get_property_value(Dev, "ID_BUS");
+    if(NULL == bus)
+    {
 
-    sEnumerateDevicesCallback callBack;
-    callBack.Data = (void*)&Paths;
-    callBack.Function = [](void* Data, const HDEVINFO& DevInfoHandle, SP_DEVINFO_DATA& DevInfoData, SP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData, eErrorCode& ErrorCode)
+        return false;
+    }
+    if( 0 == strcmp(bus, "ata") || 0 == strcmp(bus, "sata"))
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:68]: Code convention

File: StorageUtility/Linux/StorageUtility.cpp [hunk 16]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
     {
-        std::vector<String>* paths = ( std::vector<String>* )Data;
-        paths->push_back( DevDetailData->DevicePath );
-    };
+        return true;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:70]: Please replace to return(true);
Apply for the remains

File: StorageUtility/Linux/StorageUtility.cpp [hunk 17]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+    }
+    return false;
+}
 
-    error = EnumerateDevices( callBack, InterfaceClassGUID, OnErrorBehavior );
-    return( error );
-}*/
+bool IsAtaDeviceBus( sStorageAdapterProperty StorageAdapterProperty )
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:75]: update reference bypassed

File: StorageUtility/Linux/StorageUtility.cpp [hunk 18]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+{
+    if(StorageAdapterProperty.AtaBus == true)
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:77]: code convention (true should come first)

File: StorageUtility/Linux/StorageUtility.cpp [hunk 19]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-    HDEVINFO devInfoHandle =
-        SetupDiGetClassDevs
-        (
-        InterfaceClassGUID,
-        NULL,
-        NULL,
-        DIGCF_PRESENT | DIGCF_DEVICEINTERFACE
-        );
-    if ( INVALID_HANDLE_VALUE == devInfoHandle )
+    if( false == IsStorageDisk(dev))
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:84]: code convention
please fix for those remains

File: StorageUtility/Linux/Device.cpp [hunk 20]
```diff
@@ -21,84 +21,61 @@ limitations under the License.
...
 
+DeviceHandle cDevice::Handle()
+{
+    eErrorCode errorCode;
+    DeviceHandle Handle;
+    errorCode = vtStor::GetStorageDeviceHandle(this->m_DevicePath,Handle);
+    if (eErrorCode::None != errorCode)
+    {
+        // TODO:
+        // throw std::exception("Failed to create device handle", GetLastError());
```
Comment (hai-tran) [inline StorageUtility/Linux/Device.cpp:76]: What should do here?

File: StorageUtility/Linux/Device.cpp [hunk 21]
```diff
@@ -21,84 +21,61 @@ limitations under the License.
...
-    size_t devicePathLength = (tstrlen(DevDetailData->DevicePath) + 1);
-    m_DevDetailData->cbSize = DevDetailData->cbSize;
-    tmemcpy_s(m_DevDetailData->DevicePath, devicePathLength, DevDetailData->DevicePath, devicePathLength);
+void cDevice::SetDeviceHandle (const DeviceHandle Handle)
+{
+    m_DeviceHandle = Handle;
+}
+void cDevice::Data(std::unordered_map<eDeviceDataType, void*>& Data)
+{
+    // TODO:
```
Comment (hai-tran) [inline StorageUtility/Linux/Device.cpp:60]: TODO?

File: StorageUtility/Linux/Device.cpp [hunk 22]
```diff
@@ -21,84 +21,61 @@ limitations under the License.
...
 }
 
-void cDevice::DeallocateMemories()
+void cDevice::DevicePath(tchar*& DevicePath)
 {
-    HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, m_DevInfoData);
-    HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, m_DevInterfaceData);
-    HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, m_DevDetailData);
+    // TODO:
```
Comment (hai-tran) [inline StorageUtility/Linux/Device.cpp:65]: TODO?

File: StorageUtility/Linux/StorageUtility.cpp [hunk 23]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+        StorageAdapterProperty.AtaBus = false;
+        return false;
+    }
+    return true;
+}
 
-    DWORD error = 0;
-    DWORD memberIndex = 0;
-    while ( true == true )
+eErrorCode GetStorageDevicePaths( std::vector<String>& ipPaths, eOnErrorBehavior OnErrorBehavior )
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:101]: we don't need to name "ip" prefix

File: StorageUtility/Linux/Device.h [hunk 24]
```diff
@@ -33,29 +29,24 @@ namespace vtStor
...
-    void CopyMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData);
+    virtual void Data(std::unordered_map<eDeviceDataType, void*>& Data) override;
+    virtual DeviceHandle Handle() override;
+    virtual void DevicePath(tchar*& DevicePath) override;
 
-    void DeallocateMemories();*/
+    String GetDevicePath();
+    void SetDevicePath (const String DevicePath);
+    DeviceHandle GetDeviceHandle();
+    void SetDeviceHandle (const DeviceHandle Device_Handle);
```
Comment (hai-tran) [inline StorageUtility/Linux/Device.h:44]: Coding Convention: DeviceHandle

File: StorageUtility/Linux/StorageUtility.cpp [hunk 25]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
+{
+    struct udev *udev;
+    struct udev_enumerate *enumerate;
+    struct udev_list_entry *devices, *dev_list_entry;
+    struct udev_device *dev;
+
+    /* Create the udev object */
+    udev = udev_new();
+    if (NULL == udev)
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:110]: Use null_ptr

File: StorageUtility/Linux/StorageUtility.cpp [hunk 26]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-            {
-                break;
-            }
-            else
-            {
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
-        }
+        printf("ERROR: Can't create udev\n");
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:112]: Do not print anything here
Please throw an exception instead

File: StorageUtility/Linux/StorageUtility.cpp [hunk 27]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
     }
+    /* Free the enumerator object */
+    udev_enumerate_unref(enumerate);
 
-    SetupDiDestroyDeviceInfoList( devInfoHandle );
-
-    return( eErrorCode::None );
-}*/
+    udev_unref(udev);
+    return( eErrorCode::None);
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:147]: typo

File: StorageUtility/Linux/StorageUtility.cpp [hunk 28]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-                        &storageDescHeader,
-                        sizeof(STORAGE_DESCRIPTOR_HEADER),
-                        &bytesReturned,
-                        NULL
-                        );
-    if ( 0 == ret )
+    if ((Handle = open(DevicePath.c_str(), O_RDONLY|O_NONBLOCK)) < 0)
     {
-        return( eErrorCode::Io );
+        printf("ERROR: Cannot open device %s\n", DevicePath.c_str());
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:154]: Do not print anything here

File: StorageUtility/Linux/StorageUtility.h [hunk 29]
```diff
@@ -18,51 +18,27 @@ limitations under the License.
...
-        U8   Channel;
-        U8   SCSI_Id;
-        U8   Lun;
+        U8    SrbType;
+        U32   AlignmentMask;
     };
-//using HANDLE = int; add by Minh Mai
-    eErrorCode GetStorageAdapterProperty( HANDLE Handle, sStorageAdapterProperty& AdapterProperty );
+
+    eErrorCode ShowDevicePaths(const std::vector<String> Paths);
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.h:39]: Is this function just for test?

File: StorageUtility/Linux/StorageUtility.cpp [hunk 30]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-    case BusTypeiScsi:
-    {
-        return( true );
-    } break;
-    default:
-    {
-        return( false );
-    } break;
-    }*/
+    close(Handle);
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:194]: How about if close handle raised error?

File: StorageUtility/Linux/StorageUtility.cpp (hunk 5 above)
Comment (hai-tran) [inline StorageUtility/Linux/StorageUtility.cpp:32]: At the application layer, I have 2 classes that use the vtStor. One is used to issue ATA command and one is used to issue ATA and SCSI command. How about if both of them do Enumerate drive?

File: StorageUtility/Linux/StorageUtility.h [hunk 31]
```diff
@@ -18,51 +18,27 @@ limitations under the License.
...
-//    {
-//        U8    BusType;
-//        U8    SrbType;
-//        U32   AlignmentMask;
-//    };
 
     struct sStorageAdapterProperty
     {
-        //bool AtaBus;
+        bool AtaBus;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.h:33]: Update alignment

File: StorageUtility/Linux/StorageUtility.cpp [hunk 32]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-                break;
-            }
-            else
-            {
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
-        }
+        printf("ERROR: Can't create udev\n");
+        exit(1);
```
Comment (hai-tran) [inline StorageUtility/Linux/StorageUtility.cpp:113]: Why do exit here?

File: StorageUtility/Linux/StorageUtility.cpp [hunk 33]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-    if ( INVALID_HANDLE_VALUE == devInfoHandle )
+    if( false == IsStorageDisk(dev))
     {
-        return( eErrorCode::Unknown );
+        return false;
     }
+    if(true == IsAtaDeviceBus(dev))
+    {
+        StorageAdapterProperty.AtaBus = true;
+        return true;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:91]: Why put so many return on a function?

File: StorageUtility/Linux/StorageUtility.cpp [hunk 34]
```diff
@@ -17,251 +17,181 @@ limitations under the License.
...
-        {
-            if ( eOnErrorBehavior::Continue == OnErrorBehavior ) { continue; }
-            SetupDiDestroyDeviceInfoList( devInfoHandle );
-            return( eErrorCode::Memory );
-        }
-        pDevDetailData->cbSize = sizeof( SP_INTERFACE_DEVICE_DETAIL_DATA );
+    s_MappingSystemPathToProperties.clear();
+    udev_list_entry_foreach(dev_list_entry, devices)
+    {
+        const char *path;
```
Comment (tamvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:126]: code convention

**Comment by hai-tran** [comment]
Rejected because there is a lot of issues!

File: StorageUtility/Linux/StorageUtility.cpp (hunk 16 above)
Comment (minhvirtium) [inline StorageUtility/Linux/StorageUtility.cpp:70]: Why do we replace return true by return (true) ?

**Pull request closed without merging**